src/isyntax/isyntax.c

```diff
@@ -3068,12 +3073,35 @@ bool isyntax_open(isyntax_t* isyntax, const char* filename) {
 			size_t ll_coeff_block_allocator_capacity_in_blocks = block_allocator_maximum_capacity_in_blocks / 4;
 			size_t h_coeff_block_size = ll_coeff_block_size * 3;
 			size_t h_coeff_block_allocator_capacity_in_blocks = ll_coeff_block_allocator_capacity_in_blocks * 3;
-			isyntax->ll_coeff_block_allocator = block_allocator_create(ll_coeff_block_size, ll_coeff_block_allocator_capacity_in_blocks, MEGABYTES(256));
-			isyntax->h_coeff_block_allocator = block_allocator_create(h_coeff_block_size, h_coeff_block_allocator_capacity_in_blocks, MEGABYTES(256));
+			if (init_allocators) {
+				isyntax->ll_coeff_block_allocator = malloc(sizeof(block_allocator_t));
+				isyntax->h_coeff_block_allocator = malloc(sizeof(block_allocator_t));
+				*isyntax->ll_coeff_block_allocator = block_allocator_create(ll_coeff_block_size, ll_coeff_block_allocator_capacity_in_blocks, MEGABYTES(256));
+				*isyntax->h_coeff_block_allocator = block_allocator_create(h_coeff_block_size, h_coeff_block_allocator_capacity_in_blocks, MEGABYTES(256));
+				isyntax->is_block_allocator_owned = true;
+			} else {
+				// The caller must inject the allocators after return of isyntax_open().
+				isyntax->ll_coeff_block_allocator = NULL;
+				isyntax->h_coeff_block_allocator = NULL;
+				isyntax->is_block_allocator_owned = false;
+			}
 
 			success = true;
 
 			free(read_buffer);
+
+			// Populate debug info.
```

Comment:
'debug' is a bad name for it, I now rely on these 3 fields for more than debug. Need to either make them proper fields, or move to some external struct. Not sure which is best.

Reply:
Good thoughts, thank you. (That's why I did this PR).

> ... linked lists instead of arrays: is the performance impact minimal enough ... ?
I did not measure performance impact here. I chose linked list because moving elements around is O(1). The cache lists are effectively priority queues, maybe heap would be even more appropriate. But I want to leave such optimizations for later. In my latest measurements iirc, hulsken_decompress was taking >50% of runtime code - that is, doing actual useful work.

> What is the best solution for cache management
I implemented LRU cache (accounting for data dependencies), at least that was my intent modulo bugs. This should work for viewing too if you want to evict. If you don't care about releasing memory (viewing one slide at a time, this may be fine), then you don't really need an LRU cache. I think cache is orthogonal to request sequence, and the _streamer code implements mostly sequence, and _reader code implements mostly cache.

> Iterating a region ..  obfuscates .. geometric organization ... for efficient prioritizing 
I think you are assuming an access pattern, which is good when you can do that :) In ML loads for example, I have sequential access and random access both common, and those patterns won't benefit from such optimizations. But they require other optimizations - namely, cache eviction, to keep memory usage in check. As I don't really know the access pattern, LRU is the best I can do. I did an example, it should work well for linear access at the least.

Within the constructed dependency list, we can do additional optimizations, such as extra reads if beneficial, or reordering. But my assumed access pattern is "one tile at a time". Even if we were to support regions, that would bypass OpenSlide tile caching - or I will have to do complex compute of "what region corresponds to a tile". This is why I prefer tile-level access in libisyntax.

Another downside of exposing read_region as opposed to read_tile from libisyntax is that I will have to do sub-tile pixel management - cropping, concatenating, and the memory management needed for that. Tiled formats are common, and OpenSlide already handles all of that, and I assume other libraries/users would too. I agree it disables some optimizations (e.g. suppressing unnecessary neighbor decoding), but that's a tradeoff of general vs specialized. Later this is something we can control with extra flags if it is really needed.